[sig-storage] Add custom timeouts in E2E tests

bertinatto opened the pull request:
**What type of PR is this?**

/kind bug
/kind flake

**What this PR does / why we need it**:

Some storage backends can be slower or faster when it comes to certain operations like attaching or detaching a disk. In other words, what is considered to be a normal operation duration to some backends might not be OK for others.

This frequently results in flakes in certain clouds. This PR tries to improve this situation by allowing storage drivers (from E2E tests) to specify their own timeouts.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

This PR is not ready for review yet.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

bertinatto committed:
e2e: add custom timeouts to E2E framework

bertinatto committed:
e2e: use custom timeouts in all storage E2E tests

bertinatto committed:
e2e: add custom timeouts support in external storage drivers

bertinatto commented:
/sig storage

bertinatto commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -1605,6 +1606,12 @@ func (a *azureDiskDriver) PrepareTest(f *framework.Framework) (*testsuites.PerTe
+	t := framework.NewTimeoutContextWithDefaults()
+	t.PodStart = time.Minute * 10
+	return t
+}
```
This (custom timeouts in Azure Disk) is just an example of how to add custom timeouts to an in-tree driver. I plan to remove it before merging this PR.

bertinatto commented on test/e2e/framework/util.go:
```diff
@@ -71,9 +71,23 @@ import (
+	AllContainers = InitContainers | Containers | EphemeralContainers
+)
+
+// DEPRECATED constants. Use the timeouts in framework.Framework instead.
```
Ideally we'd remove these constants from here in favor of the values in framework.Framework.Timeouts. However, these consts are used in **too many places** throughout the test suite and changing all at once would make this PR too big (I tried that).

So the approach I found more convenient was to deprecate the consts in `util.go` and and remove them over time, once there isn't any test using them any more.

msau42 commented:
/assign @saikat-royc

jsafrane commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,132 @@
+	ServiceStart time.Duration
+
+	// Poll is how often to Poll pods, nodes and claims.
+	Poll time.Duration
```
Does this need to be configurable? IMO it's about the frequency and IMO we can have one that makes everyone happy - once per 2s should be OK.

bertinatto replied:
Removed

jsafrane commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,132 @@
+
+	// pv.go
+
+	// PVBindingTimeout is how long PVs have to become bound.
+	PVBinding time.Duration
+
+	// ClaimBindingTimeout is how long claims have to become bound.
+	ClaimBinding time.Duration
+
+	// PVReclaimingTimeout is how long PVs have to beome reclaimed.
+	PVReclaiming time.Duration
+
+	// PVDeletingTimeout is how long PVs have to become deleted.
+	PVDeleting time.Duration
```
These either need a better name or be merged with some timeouts above (ClaimBinding -> ClaimProvisionShort?, PVDeleting -> rename to PVDelete,...)

bertinatto replied:
These are copies of the original values [here](https://github.com/kubernetes/kubernetes/pull/96042/commits/b37e7b701afa1aeb1477b2b95bafe8fdf24e06d9#diff-f9cf7d59f38dc004bb86a00b169c989bf93d49636bdf254bf38909a3d37816e3L41-L52), but I'm OK with changing the names.

saikat-royc replied:
Let's create the new variables with proper names as Jan suggested. The old ones will eventually be deprecated.

jsafrane commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,132 @@
+import "time"
+
+// TimeoutContext contains timeout settings for several actions.
+type TimeoutContext struct {
```
There is lot of timeouts in this struct that are never used in this PR. Does it make sense to pre-allocate them for subsequent, non-storage PRs?

bertinatto replied:
My original idea was to:

1. Move all timeouts to this struct
2. Deprecate the original values used here: https://github.com/kubernetes/kubernetes/blob/f975a2d485ca7ce559d437ed6a1b07111bf1404d/test/e2e/framework/util.go#L86
3. Use the values from this struct _in all e2e tests_, not only storage tests.

However, when I got to step 3 above I noticed that this PR would become excessively big, so I started with storage tests only and hoped that eventually the other tests would use these values instead (since the values in utils.go are deprecated).

Does this make sense or should I just add to this struct the timeouts I'm going to use in storage tests?

The reason I wanted to do this is because currently timeout values are scattered around framework package: utils.go, pv/pv.go etc. So I'd like this file to be the source of truth for all values.

saikat-royc replied:
Keeping the timeouts in a single place (timeouts.go) makes the code readable. I think we can start with the timeouts we are using in storage tests.

bertinatto replied:
Sounds good. Removed everything that was not used.

bertinatto commented:
/retest

saikat-royc commented on test/e2e/storage/external/external.go:
```diff
@@ -302,6 +309,34 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(e2econfig *testsuites
 
+func (d *driverDefinition) GetTimeouts() *framework.TimeoutContext {
+	timeouts := framework.NewTimeoutContextWithDefaults()
+	if d.Timeouts != nil {
```
Can we instead check if d.Timeouts == nil { return timeouts} and avoid the indent.

saikat-royc commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,124 @@
+// NewTimeoutContextWithDefaults returns a TimeoutContext with default values.
+func NewTimeoutContextWithDefaults() *TimeoutContext {
+	return &TimeoutContext{
+		PodList:                 1 * time.Minute,
```
Don't think we need custom timeouts for listing/getting resources. Thoughts?

saikat-royc commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,124 @@
+
+// NewTimeoutContextWithDefaults returns a TimeoutContext with default values.
+func NewTimeoutContextWithDefaults() *TimeoutContext {
+	return &TimeoutContext{
```
Can we create const for the default timeout values in this file? and then use them here.
e.g PVBindingTimeout etc

saikat-royc commented on test/e2e/storage/external/external.go:
```diff
@@ -128,6 +131,8 @@ type driverDefinition struct {
 	// use topology to ensure that pods land on the right node(s).
 	ClientNodeName string
+
+	Timeouts map[string]string
```
Can you put some comments for this map

saikat-royc commented on test/e2e/storage/external/external.go:
```diff
@@ -302,6 +309,34 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(e2econfig *testsuites
+			if err != nil {
+				// We can't use ExpectNoError() because his method can be called out of an It(),
+				// so we simply log the error and return the default timeouts.
+				klog.Errorf("Could not parse duration: %v", err)
```
Can you also specify the key here for debugging: "Could not parse duration %v for key %s"

saikat-royc commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,124 @@
+		RestartPodReadyAgain:    5 * time.Minute,
+		SnapshotCreate:          5 * time.Minute,
+		SnapshotDelete:          5 * time.Minute,
+		// pv.go
```
align the indentation values

saikat-royc commented:
@bertinatto Thanks for making the changes. Can you rebase the patch and add the changes to new test suites like [fsgroupchangepolicy.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/fsgroupchangepolicy.go)

bertinatto commented on test/e2e/framework/pv/pv.go:
```diff
@@ -38,18 +39,6 @@ const (
-	PVReclaimingTimeout = 3 * time.Minute
-
-	// PVDeletingTimeout is how long PVs have to become deleted.
-	PVDeletingTimeout = 3 * time.Minute
```
These values are not used anywhere else, so I removed instead of deprecating them.

saikat-royc commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,90 @@
+	podStartTimeout            = 5 * time.Minute
+	podStartShortTimeout       = 2 * time.Minute
+	podDeleteTimeout           = 5 * time.Minute
+	claimProvisionTimeout      = 5 * time.Minute
```
minor nit: its more readable to keep the timeouts for a given resource together. e.g claimProvisionTimeout, claimProvisionShortTimeout, claimBoundTimeout

saikat-royc commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,90 @@
+
+	// PodStartShort is same as `PodStart` to wait for the pod to be started, but shorter.
+	// Use it case by case when we are sure pod start will not be delayed.
+	// minutes by slow docker pulls or something else.
```
please revisit this sentence. "minutes by slow docker pulls or something else"

saikat-royc commented on test/e2e/framework/timeouts.go:
```diff
@@ -0,0 +1,90 @@
+		PodStart:            podStartTimeout,
+		PodStartShort:       podStartShortTimeout,
+		PodDelete:           podDeleteTimeout,
+		ClaimProvision:      claimProvisionTimeout,
```
same comment as above about grouping resources together.

saikat-royc commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -383,7 +384,7 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
-			err = e2eevents.WaitTimeoutForEvent(m.cs, pod.Namespace, eventSelector, msg, framework.PodStartTimeout)
+			err = e2eevents.WaitTimeoutForEvent(m.cs, pod.Namespace, eventSelector, msg, f.Timeouts.PodStart)
 			if err != nil {
 				podErr := e2epod.WaitTimeoutForPodRunningInNamespace(m.cs, pod.Name, pod.Namespace, 10*time.Second)
```
10*time.Second replace this with podStartShortTimeout?

saikat-royc commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -45,6 +45,7 @@ import (
 // Not all parameters are used by all tests.
```
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/provisioning.go#L388,
```framework.ExpectNoError(e2epv.WaitForPersistentVolumeDeleted(client, pv.Name, 5*time.Second, 20*time.Minute))```
there is a wait for pv delete step that has 20 mins timeout not sure why. This needs to be revisited, perhaps in a different PR.

msau42 replied:
I didn't add the comment, I just moved it to a new place :-)

I had to trace back quite a bit to where this comment was originally added: https://github.com/kubernetes/kubernetes/commit/46fb4172999bebcf822a2953e1783ed0428adc70

I think the 20 mins is to also account for the time period that it may take to unmount and detach the volume. 20 minutes seems quite long and could probably be part of the customizable timeouts?

saikat-royc commented on test/e2e/framework/volume/fixtures.go:
```diff
@@ -441,11 +441,11 @@ func runVolumeTesterPod(client clientset.Interface, config TestConfig, podSuffix
 	if slow {
 		err = e2epod.WaitForPodRunningInNamespaceSlow(client, clientPod.Name, clientPod.Namespace)
```
this function indirectly uses https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pod/wait.go where the pod specific timeouts are defined. Those timeouts should also eventually move to the timeouts.go. we can put a todo here?

saikat-royc commented:
/cc @jingxu97 for windows specific tests

jingxu97 commented on test/e2e/storage/external/external.go:
```diff
@@ -302,6 +312,42 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(e2econfig *testsuites
+
+	// Override the default timeouts with the custom ones
+	err = json.Unmarshal(t, &timeouts)
+	if err != nil {
```
just want to make sure at this point, the returned timeouts will still be the default value?
also in the log, maybe mention returning default timeout?

bertinatto replied:
Right, `timeouts` contains the default values at this point.

Extended log message...

jingxu97 commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -129,6 +129,19 @@ type SnapshottableTestDriver interface {
 }
 
+// CustomTimeoutsTestDriver represents an interface fo a TestDriver that supports custom timeouts.
+type CustomTimeoutsTestDriver interface {
```
just a suggestion: TestDriverWithTimeout?

bertinatto replied:
That sounds better indeed, but it'd differ from the other interfaces (they are suffixed with `TestDriver`)

jingxu97 commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -129,6 +129,19 @@ type SnapshottableTestDriver interface {
+	GetTimeouts() *framework.TimeoutContext
+}
+
+func getCustomTimeouts(driver TestDriver) *framework.TimeoutContext {
```
getDriverTimeouts?

bertinatto replied:
Renamed

jingxu97 commented:
thanks for the PR! @bertinatto just a few small comments. overall lgtm.

cc @Jiawei0227

saikat-royc commented on test/e2e/storage/external/external.go:
```diff
@@ -302,6 +312,42 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(e2econfig *testsuites
+	}
+
+	// Use a temporary map to hold the timeouts specified in the manifest file
+	c := make(map[string]time.Duration)
```
The map should be created with default values. And overwrite the values for the keys provided in the driver config.

bertinatto replied:
The `timeouts` struct above contains the default timeouts. This map only contains the custom values provided by the driver config.

Note that down below map is then merged with the struct, overwriting the default values with the custom ones.

saikat-royc replied:
Ah I see. thanks for pointing it out. missed it.

Jiawei0227 commented:
/retest

msau42 commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -383,7 +384,7 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 			msg := "AttachVolume.Attach failed for volume"
 
-			err = e2eevents.WaitTimeoutForEvent(m.cs, pod.Namespace, eventSelector, msg, framework.PodStartTimeout)
+			err = e2eevents.WaitTimeoutForEvent(m.cs, pod.Namespace, eventSelector, msg, f.Timeouts.PodStart)
```
Hm so if a plugin decides to set a non-default timeout, then that will impact the timeout for non driver specific tests?

Is that what we want?

saikat-royc replied:
can you point to an e.g for non-driver specific test?

msau42 replied:
I was thinking specifically about this csi mock driver test. If you run the csi mock driver test along with a cloud plugin that takes minutes to provision/attach, etc, then that increases the timeout on the mock driver tests, which aren't expected to take very long, and could mask other problems in the system.

saikat-royc replied:
I see, this seems to be a csi mock plugin specific problem. For the testsuites, they are run in the context of a driver (intree, csi or external). We can remove the non-default timeouts for the mock driver. thoughts?

msau42 replied:
I guess right now these timeouts are only configurable via external storage interface or a custom ginkgo suite, so it's unlikely for someone to be running both csi mock and external storage tests in the same run. I would be more concerned if we want to extend this to all e2es in the future.

bertinatto replied:
Notice that this csi mock driver test is not using custom timeouts, it's using the default ones:

https://github.com/kubernetes/kubernetes/pull/96042/files/f594e0f7737a0d281e9b08ddf1b1ec770755ee6a#diff-ffda3f0c41f167e7deedad89b23cf9b89d0e9805f910c81e075baa19d667f21fR137

Even if it was using the custom timeouts, it would get the values returned by the mock driver's `GetTimeouts()` method, as in:

```diff
diff --git a/test/e2e/storage/drivers/csi.go b/test/e2e/storage/drivers/csi.go
index 03ec35e557c..d4956d47389 100644
--- a/test/e2e/storage/drivers/csi.go
+++ b/test/e2e/storage/drivers/csi.go
@@ -278,6 +278,7 @@ type CSIMockDriverOpts struct {
 var _ testsuites.TestDriver = &mockCSIDriver{}
 var _ testsuites.DynamicPVTestDriver = &mockCSIDriver{}
 var _ testsuites.SnapshottableTestDriver = &mockCSIDriver{}
+var _ testsuites.CustomTimeoutsTestDriver = &mockCSIDriver{}

 // InitMockCSIDriver returns a mockCSIDriver that implements TestDriver interface
 func InitMockCSIDriver(driverOpts CSIMockDriverOpts) testsuites.TestDriver {
@@ -336,6 +337,12 @@ func InitMockCSIDriver(driverOpts CSIMockDriverOpts) testsuites.TestDriver {
        }
 }

+func (m *mockCSIDriver) GetTimeouts() *framework.TimeoutContext {
+       timeouts := framework.NewTimeoutContextWithDefaults()
+       timeouts.PodStart = time.Second * 5
+       return timeouts
+}
+
 func (m *mockCSIDriver) GetDriverInfo() *testsuites.DriverInfo {
        return &m.driverInfo
 }
```

msau42 replied:
Awesome, this addresses my concern for now thanks!

saikat-royc commented:
Looking at one more example of subpath.go tests, the function [testReadFile](
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L682:6), uses [TestContainerOutput](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L687). The calls lands finally to [MatchContainerOutput](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L803:21). This means for a driver which supports waitforfirstcustomer  (delayed binding), we are not able to control the pod start time, since the start time would encompass the volume provisioning time too.

bertinatto committed:
e2e: add new custom timeout value for slow PV deletion

bertinatto committed:
e2e: promote use of functions that support custom timeouts in storage tests.

WaitForPodSuccessInNamespace[Slow] are replaced by WaitForPodSuccessInNamespaceTimeout(),
so that custom timeouts are used instead of the hardcoded ones.

msau42 commented:
/approve

Will let @saikat-royc give lgtm

bertinatto commented:
/retest

saikat-royc commented on test/e2e/storage/pvc_protection.go:
```diff
@@ -18,12 +18,13 @@ package storage
 	"time"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
is this change needed?

saikat-royc commented on test/e2e/storage/vsphere/vsphere_stress.go:
```diff
@@ -23,7 +23,7 @@ import (
 	"github.com/onsi/ginkgo"
 	"github.com/onsi/gomega"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
same comment here

saikat-royc commented:
The changes look good to me with 2 minor comments. 

/lgtm

saikat-royc commented:
@bertinatto  found some more instances of WaitForPodSuccessInNamespace() calls in the storage testsuites. e.g [subpath.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L1013).
some more in provisioning.go etc, where we need to use the configurable pod start

saikat-royc commented:
/lgtm cancel

bertinatto reviewed (commented):
```
@bertinatto found some more instances of WaitForPodSuccessInNamespace() calls in the storage testsuites. e.g subpath.go.
some more in provisioning.go etc, where we need to use the configurable pod start
```
@saikat-royc I think I've replaced all occurrences of  `WaitForPodSuccessInNamespace*` by `WaitForPodSuccessInNamespaceTimeout` in storage tests. PTAL

bertinatto commented:
/retest

saikat-royc commented:
/lgtm

saikat-royc commented:
/retest

saikat-royc commented:
/retest

msau42 commented:
/priority important-soon
/triage accepted

saikat-royc commented:
The test `pull-kubernetes-e2e-gce-ubuntu-containerd` seems to be failing due to failure to pull image
```BackOff: Back-off pulling image "gcr.io/k8s-authenticated-test/agnhost:2.6"```

bertinatto commented:
/retest

bertinatto commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

bertinatto commented:
```
Kubernetes e2e suite: [sig-apps] ReplicationController should serve a basic image on each replica with a private image expand_more 	5m10s
Kubernetes e2e suite: [sig-apps] ReplicaSet should serve a basic image on each replica with a private image expand_more
```

The failure doesn't seem related to this PR.

/retest

bertinatto commented:
xref: https://github.com/kubernetes/kubernetes/issues/97002
